### ISSUE TITLE
Build Environment (Docker 1.9)

### ISSUE THREAD

**Issue by nzoschke** [convox/rack#156]
Docker 1.9 builder introduces build time arguments:

http://docs.docker.com/engine/reference/builder/#arg

With Docker 1.9 installed `convox start` supports defaults and pass through build arguments locally, but `convox build` runs with an empty environment. The ideal platform would allow a user to:
- Define ARG FOO= defaults in Dockerfile for `convox start`
- `convox env:set FOO=bar --build` for an pp
- See `docker build --build-arg FOO=bar` in `convox build`

This will open up powerful dev / build / prod workflows such as:
- Run a rails app locally with dynamic assets
- Build a rails app on convox with assets precompile and S3 assets sync
- Run a rails app in production with assets served from S3

**Comment by ddollar** [comment]
Development probably needs separate build env outside of the Dockerfile also if there are any secrets going on here.

**Comment by xtfxme** [comment]
"If a user specifies a build argument that was not defined in the Dockerfile, the build outputs an error: One or more build-args were not consumed, failing build."

I think you'd have to enumerate the `ARG` lines from the target Dockerfile to make it work like that (is there an existing programmatic way to analyze a Dockerfile?). I'm not sure you want to conflate `ENV` args with build `ARG`, they are different things in docker with different semantics.

If convox adds support for version 2 of `docker-compose.yml`, there is a dedicated section for build args: https://docs.docker.com/compose/compose-file/#args

I do see some appeal in reusing runtime `ENV` args during the build, but we should probably spend time considering the ramifications, and why docker doesn't do this. Eg. you would not want a build-time `http_proxy` or similar meant to speed up builds by using local proxy-caches (say apt-get) exported into the actual container at runtime, in production.

**Comment by xtfxme** [comment]
It sounds like this issue's text might agree with what I just said, so if that's the case, carry on... nothing to see here.

**Comment by mattmanning** [comment]
Just for more context here are a couple of use cases for build args:

- Ruby Bundler: when building for local development you probably want to run `bundle install`, but for production you probably want to do something like `bundle install --without test development`. There are probably also system level packages that you could avoid installing for production, for example the `libqt4` packages needed for the `capybara-webkit` testing library.

- Meteor: For local development the app can be booted simply with `meteor run`, but in production you want to go through `meteor build` which compiles a node app.

**Comment by cayblood** [comment]
We have a phoenix app that reads from env vars at compile (build) time and caches the results in the build, which makes it difficult/impossible for us to use dummy values during builds. It would save us a lot of grief if we could somehow use the env vars we have set up in the convox console during build/deploy time.

**Comment by mattmanning** [comment]
Proposed design

**CLI:**
`convox args` command with `get|set|unset` subcommands exactly like convox env. Any values set this way override values in docker-compose.yml.

**docker-compose.yml:**
Support Compose v2 format build args. https://docs.docker.com/compose/compose-file/#args

Local dev:
An optional `.args` file. Any values in this file override those in `docker-compose.yml`.

**Comment by frankleng** [comment]
similar issue to @cayblood 
we have build time env vars that define the version of db's docker image to pull, and commands that are env-specific to create the db cluster.
Would ideally store these as build-time vars instead of creating a bunch of static docker-compose.yml files.

**Comment by frankleng** [comment]
@mattmanning like your proposal.
also this may be outside the scope of this issue - how about storing these as secure vars via console.convox.com? only accessible to devs with devops access.

**Comment by mwarkentin** [comment]
We have a private pypi server we use for installing python packages - I think something like this would be helpful so that we don't need to hardcode the credentials in the Dockerfile.

**Comment by dlanger** [comment]
Similar to the mention of `bundler` above, we have different `requirements.txt` files for production vs. development, and we'd use this with our CI to get the right packages installed.

**Comment by mattmanning** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;0alJ6y5j&#x2F;82-build-arguments"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> build arguments</a>

**Comment by mattmanning** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;0alJ6y5j&#x2F;82-build-arguments"><img src="https:&#x2F;&#x2F;github.trello.services&#x2F;images&#x2F;trello-icon.png" width="12" height="12"> build arguments</a>

**Comment by cayblood** [comment]
@mattmanning FYI this link is invisible to the public.

**Comment by leftdevel** [comment]
Hi guys. Sorry for comment-bombing. Is there any workaround for passing build args with convox? Currently it ignores them in the docker-compose.yml file : (

**Comment by joshk** [comment]
I would very very much like this feature! We use Sidekiq pro, which uses a private gem source, and we don't want to commit the auth details to our Gemfile, so instead use an Env var.

This feature would allow us to keep the same security policy we have right now.

**Comment by frankleng** [comment]
@mattmanning any movement on this issue yet?

**Comment by sc-chad** [comment]
This is something that would definitely save me writing some sed substitution wrappers or a bunch of static docker-compose files.

My use case is we use puppet for our configuration management.  The main thing puppet does is build .env files dynamically for our rails apps based on some metadata.  Prior to docker I was using ec2-tags as the metadata that defined what the instance was.  This let me populate the .env file dynamically with puppet.

Now in docker I'm doing pretty much the same thing, but I can't query anything dynamically to define what the container is supposed to be.

Rough Example:
**docker-compose.yml**
```
version: "2"
services:
  workers:
    args:
      - product
      - role
      - environment
      - client
    build:
      context: .
    command: bundle exec rake some_worker
    environment:
      - product=rais_worker01
      - role=worker
      - environment=development
      - client=some_client
```
**Dockerfile**

```
# This is built from the amazon-linux-image docker directory
FROM amazon-linux-image/main:latest

ARG environment
ARG product
ARG role
ARG client

# run bundle install
COPY Gemfile      /app/Gemfile
COPY Gemfile.lock /app/Gemfile.lock
RUN su root -l -c "cd /app && bundle install --path=vendor/bundle --jobs 4 --retry 3"

# echo out environment variables
RUN echo "Environment: $environment"
RUN echo "Product: $product"
RUN echo "Role: $role"
RUN echo "Client: $client"

# runs puppet to configure the services / configuration files
RUN environment=$environment \
    product=$product \
    role=$role \
    client=$client \
    /opt/puppetlabs/bin/puppet apply --hiera_config=/opt/puppet_core/hiera.yaml --modulepath=/opt/puppet_modules/modules --environmentpath=/opt/puppet_modules /opt/puppet_modules/manifests/site.pp

# copy the rest of the app
COPY . /app
```

**Convox Start Command**
`convox start --args environment=production,product=rails_worker_02,role=worker,client=someclient`

This would build my .env file with the correct secrets and variables for this environment prior to running the `bundle exec rake some_worker` from docker-compose.yml.  This also lets me programmatically build and test locally the same way it's used in AWS.

**Comment by nzoschke** [comment]
This shipped in #157 #1666
